cmd/influx/cli/cli.go

```diff
@@ -777,38 +778,103 @@ func (c *CommandLine) writeJSON(response *client.Response, w io.Writer) {
 	fmt.Fprintln(w, string(data))
 }
 
+func tagsEqual(prev, current map[string]string) bool {
+	if prev == nil && current == nil {
+		return true
+	}
+	if len(prev) != len(current) {
+		return false
+	}
+	return reflect.DeepEqual(prev, current)
+}
+
+func columnsEqual(prev, current []string) bool {
+	if prev == nil && current == nil {
+		return true
+	}
+	if len(prev) != len(current) {
+		return false
+	}
+	return reflect.DeepEqual(prev, current)
+}
+
+func headersEqual(prev, current models.Row) bool {
+	if prev.Name != current.Name {
+		return false
+	}
+	if prev.Tags == nil && current.Tags == nil {
+	}
+	if !tagsEqual(prev.Tags, current.Tags) {
+		return false
+	}
+	if !columnsEqual(prev.Columns, current.Columns) {
+		return false
+	}
+
+	return true
+}
+
 func (c *CommandLine) writeCSV(response *client.Response, w io.Writer) {
 	csvw := csv.NewWriter(w)
+	var previousHeaders models.Row
 	for _, result := range response.Results {
+		suppressHeaders := len(result.Series) > 0 && headersEqual(previousHeaders, result.Series[0])
+		if !suppressHeaders && len(result.Series) > 0 {
+			previousHeaders = models.Row{
+				Name:    result.Series[0].Name,
+				Tags:    result.Series[0].Tags,
+				Columns: result.Series[0].Columns,
+			}
+		}
+
 		// Create a tabbed writer for each result as they won't always line up
-		rows := c.formatResults(result, "\t")
+		rows := c.formatResults(result, "\t", suppressHeaders)
 		for _, r := range rows {
 			csvw.Write(strings.Split(r, "\t"))
 		}
-		csvw.Flush()
 	}
+	csvw.Flush()
```

Comment:
You only need to call flush once.  It's always streaming.  And if you call flush to early, it actually messes up the alignment on the output (didn't read the source, but assuming it resets the column width tracking).  There is no benefit to calling flush more than once regardless.